_posts/2020-08-11-relay-pagingation-containers-demystified.markdown

```diff
@@ -0,0 +1,234 @@
+      /* Render each artwork */
+      <Text key={node.id}>{node.name}</Text>
+    )}
+    {this.props.relay.hasMore &&
+      <Button onPress={() => this.props.relay.loadMore() } text="Load next page" />
+    )
+  }
+}
+
+export ArtistFragmentContainer = createPaginationContainer(Artist, {
+  artist: graphql`
+    fragment Artist_artist on Artist @argumentDefinitions(
+      count: { type: "Int", defaultValue: 10 }
+      cursor: { type: "String" } {
+      name
+      id
+      artworksConnection (first: $count, after: $cursor) @connection(key: "Artist_artworks") {
+        edges {
+          node {
+            title # Now fetch all the artwork data
+            id
+          }
+        }
+      }
+    }
+  `
+}, {
+  direction: "forward",
+  getConnectionFromProps(props) {
+    return props.artist.artworks
+  },
+  getFragmentVariables(prevVars, count) {
+    return {
+      ...prevVars,
+      count,
+    }
+  },
+  getVariables(props, { count, cursor }, fragmentVariables) {
+    return {
+      id: props.artist.id,
+      count,
+      cursor,
+    }
+  },
+  query: graphql` # Here is the query to fetch any specific page
+    query ArtistArtworksQuery(
+      $id: ID!
+      $count: Int!
+      $cursor: String) {
+      artist(id: $id) {
+        ...Artist_artist @arguments(
+          count: $count
+          cursor: $cursor
+        )
+      }
+    }
+  `
+})
+```
+
+Wow, that's a lot! I don't want to get too bogged down in details, so let's break this apart at a high level:
+
+- We changed the React component to show a list of artworks and include a button to load the next page.
+- We changed from `createFragmentContainer` to using `createPaginationContainer`.
+- We added GraphQL fragment variables for `count` and `cursor` to be passed through to the new
+  `artworksConnection`, which we also added.
+- Finally, we added a whole new configuration parameter to `createPaginationContainer`.
+
+This last bit is the part where I see the most frustration. Hopefully what follows will clear things up.
+
+I like to always start by [reading the docs](https://relay.dev/docs/en/pagination-container). The `direction` key
+is the direction that we paginate through, either `"forward"` or `"backward"`. `getConnectionFromProps` is a
+function that returns the GraphQL connection, in case the query has more than one. And `query` is used to fetch any
+specific page of results.
+
+Those all makes sense to me, but then we arrive at the real gotchas: `getFragmentVariables` and `getVariables`. The
+docs are helpful, but only if you understand
+[the internals of how Relay works](https://relay.dev/docs/en/runtime-architecture.html). Relay has a sophisticated
+architecture that delivers some really well-performing code, but its abstractions sometimes
+"[leak](https://en.wikipedia.org/wiki/Leaky_abstraction)" and you have to deal with implementation details that
+Relay doesn't really want you to even be aware of (in this case, the Relay store).
```

Comment:
> you have to deal with implementation details that Relay doesn't really want you to even be aware of (in this case, the Relay store).

Curious, what makes you say that?

Reply:
Yeah, I waffled on this wording. More accurately, Relay doesn't want you using store directly but rather, as the docs point out, to build abstractions on top of instead. I will rephrase 👍